### PR TITLE
[Core] Special Transactions base layer

### PR DESCRIPTION
**Pull request by random-zebra**
Introduce the concepts of "transaction type" and "extra payload" inspired by Dash's [DIP2](https://github.com/dashpay/dips/blob/master/dip-0002.md) and integrate them in our customized serialization for SAPLING_VERSION transactions.

Even though we won't probably activate any special type with v5.0, having the new serialization already in place allows us to enable them in a future upgrade, without needing to increment the transaction version again.

This PR includes basic non-contextual validation rules (that can only be safely enforced after Sapling NU):
- nVersion=1 supports only nType=0
- A tx can have extra payload only if nType!=0 (thus, due to the previous rule, only if nVersion>=2)
- A tx with nType!=0 cannot have empty payload.
- The serialized payload cannot be bigger than a fixed max size (valid for all payload types, currently set at 10 kB).

This also prepares the method stubs for the specific per-payload-type validation rules, and for the tiertwo data processing, that will be added in the future.

### CODE REVIEW

**Review by furszy — COMMENTED** [review]
code review c1061e66 👍 .

Initial comment, need to add the `extraPayload` member to the `CMutableTransaction` class, the `CMutableTransaction(const CTransaction& tx)` constructor and serialization. Otherwise special txs cannot be created for testing purposes (at the end, `CTransaction` is an immutable class).

File: src/tiertwo/specialtx_validation.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+    // After Sapling activation.
+    // v1 can only be Type=0
+    if (!tx.isSaplingVersion() && tx.nType != CTransaction::TxType::NORMAL) {
+        return state.DoS(100, error("%s: Type %d not supported with version 0", __func__, tx.nType),
+                         REJECT_INVALID, "bad-txns-type-version");
+    }
+    if (tx.nType == CTransaction::TxType::NORMAL) {
+        // Type-0 txes don't have extra payload
+        if (hasExtraPayload) {
+            return state.DoS(100, error("%s: Type 0 desn't support extra payload", __func__),
```
Comment (furszy) [inline src/tiertwo/specialtx_validation.cpp:34]: typo "desn't"/"doesn't

**Comment by random-zebra** [comment]
Rebased on master, and cherry-picked a couple commits from @furszy, adding a unit-test for the basic validation rules of special transactions.

**Review by furszy — APPROVED** [review]
utACK b6d7bc3 after rebase, nice work.

There is a commit with an [SQUASH] label that could be squashed.

**Comment by random-zebra** [comment]
Rebased on master and commit squashed.

**Review by furszy — APPROVED** [review]
utACK 2fbb932 after rebase.

**Comment by furszy** [comment]
testnet is getting closer, merging..

**Pull request merged**